libcst/codemod/visitors/_gather_exports.py

```diff
@@ -43,68 +46,86 @@ def __init__(self, context: CodemodContext) -> None:
-            self.explicit_exported_objects.add(string.evaluated_value)
-
-        # Don't need to visit children
+    def _handle_assign_target(
+        self, target: cst.BaseExpression, value: cst.BaseExpression
+    ) -> bool:
+        target_name = get_full_name_for_node(target)
+        if target_name == "__all__":
+            # Assignments such as `__all__ = ["os"]`
+            # or `__all__ = exports = ["os"]`
+            if isinstance(value, (cst.List, cst.Tuple, cst.Set)):
+                self._is_assigned_export.add(value)
+                return True
+        elif isinstance(target, cst.Tuple) and isinstance(value, cst.Tuple):
+            # Assignments such as `__all__, x = ["os"], []`
+            for element_idx, element_node in enumerate(target.elements):
+                element_name = get_full_name_for_node(element_node.value)
+                if element_name == "__all__":
+                    element_value = value.elements[element_idx].value
+                    if isinstance(element_value, (cst.List, cst.Tuple, cst.Set)):
+                        self._is_assigned_export.add(element_value)
+                        return True
+        return False
+
+    def visit_List(self, node: cst.List) -> bool:
+        if node in self._is_assigned_export:
+            self._in_assigned_export.add(node)
+        return bool(self._in_assigned_export)
+
+    def leave_List(self, original_node: cst.List) -> None:
+        self._is_assigned_export.discard(original_node)
+        self._in_assigned_export.discard(original_node)
+
+    def visit_Tuple(self, node: cst.Tuple) -> bool:
+        if node in self._is_assigned_export:
+            self._in_assigned_export.add(node)
+        return bool(self._in_assigned_export)
+
+    def leave_Tuple(self, original_node: cst.Tuple) -> None:
+        self._is_assigned_export.discard(original_node)
+        self._in_assigned_export.discard(original_node)
+
+    def visit_Set(self, node: cst.Set) -> bool:
+        if node in self._is_assigned_export:
+            self._in_assigned_export.add(node)
+        return bool(self._in_assigned_export)
+
+    def leave_Set(self, original_node: cst.Set) -> None:
+        self._is_assigned_export.discard(original_node)
+        self._in_assigned_export.discard(original_node)
+
+    def visit_SimpleString(self, node: cst.SimpleString) -> bool:
+        self._handle_string_export(node)
+        return False
+
+    def visit_ConcatenatedString(self, node: cst.ConcatenatedString) -> bool:
+        self._handle_string_export(node)
         return False
+
+    def _handle_string_export(
+        self, node: Union[cst.SimpleString, cst.ConcatenatedString]
+    ) -> None:
+        if self._in_assigned_export:
+            name = node.evaluated_value
+            if name is None:
+                return
+            scope = self.get_metadata(cst.metadata.ScopeProvider, node, None)
+            if scope:
+                defined = name in scope.assignments
```

Comment:
This will be `False` in a scenario like
```
from .submodule import *
__all__ = ["SomethingFromSubmodule"]
```

Reply:
The reason I bring this up is because star imports are more common in cases when `__all__` is defined, which is precisely the case this visitor is useful in.

Reply:
That's too bad ...maybe we want to just not set those to `False` if there are star imports... there is value in knowing exported symbols are undefined in the other cases